xla/service/collective_quantizer.cc

```diff
@@ -50,14 +60,121 @@ auto ConvertToNarrowerType() {
   return m::Convert().WithPredicate(converts_to_narrower_type);
 }
 
+// Matches a type conversion to a type with a larger byte size than that of the
+// operand.
+auto ConvertToWiderType() {
+  auto converts_to_wider_type = [](const HloInstruction* instr) -> bool {
+    return ShapeUtil::ByteSizeOfPrimitiveType(instr->shape().element_type()) >
+           ShapeUtil::ByteSizeOfPrimitiveType(
+               instr->operand(0)->shape().element_type());
+  };
+  return m::Convert().WithPredicate(converts_to_wider_type);
+}
+
+bool IsSupportedCollective(HloInstruction* instr) {
+  return instr->operand_count() == 1 &&
+         (instr->opcode() == HloOpcode::kAllGather ||
+          instr->opcode() == HloOpcode::kAllToAll ||
+          instr->opcode() == HloOpcode::kCollectiveBroadcast ||
+          instr->opcode() == HloOpcode::kCollectivePermute);
+}
+
+// Sequentially applies the ops in unaries to the output of instr.
+HloInstruction* ApplyUnary(HloInstruction* instr,
+                           const std::vector<HloInstruction*>& unaries) {
+  for (HloInstruction* unary_op : unaries) {
+    instr = instr->AddInstruction(unary_op->CloneWithNewOperands(
+        ShapeUtil::MakeShapeWithDenseLayout(
+            instr->shape().element_type(), unary_op->shape().dimensions(),
+            unary_op->shape().layout().minor_to_major()),
+        {instr}));
+  }
+  return instr;
+}
+
+// Recursively collects unary, divide, or multiply operands of instr until a
+// conversion to a wider type is reached. Returns an empty vector when no
+// conversion is reached.
+std::vector<HloInstruction*> FindDequantizationSubgraphRecursive(
+    HloInstruction* instr, absl::flat_hash_set<int>& visited_instrs,
+    std::vector<HloInstruction*> subgraph) {
+  // Avoid visiting the same instruction more than once.
+  if (!visited_instrs.emplace(instr->unique_id()).second) {
+    return {};
+  }
+
+  subgraph.emplace_back(instr);
+  if (Match(instr, ConvertToWiderType())) {
+    return subgraph;
+  }
+  if (instr->operand_count() == 1 || instr->opcode() == HloOpcode::kDivide) {
+    return FindDequantizationSubgraphRecursive(instr->mutable_operand(0),
+                                               visited_instrs, subgraph);
+  } else if (instr->opcode() == HloOpcode::kMultiply) {
+    for (HloInstruction* operand : instr->unique_operands()) {
+      auto binary_subgraph = FindDequantizationSubgraphRecursive(
+          operand, visited_instrs, subgraph);
+      if (!binary_subgraph.empty()) {
+        return binary_subgraph;
+      }
+    }
+  }
+  return {};
+}
+
+// Returns true iff instr describes a dequantization, i.e. a multiplication or
+// division by a broadcasted scalar operating on a type conversion, or a plain
+// type conversion to a wider type. Unary bitcast, copy, reshape or slice ops
+// may follow the dequantization or type conversion.
```

Comment:
Be more explicit about the fact that `instr` itself can refer to a unary op. E.g. by replacing the last sentence with

Also returns true if instr refers to the last instruction of a sequence of unary bitcast, copy, reshape or slice ops preceded by a dequantization.